File: 이승민/[20.03.30]2178.cpp
#include <iostream>
#include <queue>

using namespace std;
typedef pair<int, int> p;
queue<p> loc;

int n = 0, m = 0, x = 0, y = 0, nX = 0, nY = 0;
int maze[101][101], dx[4] = {0, 1, 0, -1}, dy[4] = {1, 0, -1, 0};

void bfs(){
    while( !loc.empty() ){
        p x_y = loc.front();
        loc.pop();
        
        y = x_y.first;
        x = x_y.second;

        for(int i = 0; i < 4; i++){
            nX = x + dx[i];
            nY = y + dy[i];

            if( nX < 0 || nX >= m || nY < 0 || nY >= n) continue;
            if( maze[nY][nX] != 1) continue;

            loc.push({nY, nX});
            maze[nY][nX] = maze[y][x] + 1;
        }
    }
}

int main(){
    
    // 세로 가로
    scanf("%d %d", &n, &m);
    
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            scanf("%1d", &maze[i][j]);
        }
    }
    
    loc.push({0,0});
    bfs();

    printf("%d", maze[n-1][m-1]);

    return 0;
}

File: 이승민/[20.03.30]2178.py
from collections import deque

def bfs():
    dx = [0,1,0,-1]
    dy = [1,0,-1,0]

    while loc:
        x_y = loc.popleft()
        y, x = x_y[0], x_y[1]

        for i in range(4):
            nX = x + dx[i]
            nY = y + dy[i]

            if nX < 0 or nX >= m or nY < 0 or nY >= n: continue
            if maze[nY][nX] != 1: continue

            loc.append((nY, nX))
            maze[nY][nX] = maze[y][x] + 1

if __name__ == "__main__":

    n, m = map(int, input().split())
    maze = [[int(i) for i in input()] for j in range(n)]

    loc = deque([])
    loc.append((0,0))

    bfs()

    print(maze[n-1][m-1])
